File: C++/1269-Number of Ways to Stay in the Same Place After Some Steps.cpp
class Solution {
public:
    int steps;
    long kMod = 1e9 + 7;
    int arrLen;
    vector<vector<int>> memo;
    
    int dp(int pos, int step){
        if (step == steps){
            if (pos == 0) return 1;
            else return 0;
        }
        if (memo[pos][step] != -1) return memo[pos][step];
        
        // stay
        int ways = dp(pos, step+1);
        // left
        if (pos > 0)
            ways = (ways + dp(pos-1, step+1))%kMod;
        // right
        if (pos < arrLen - 1){
            ways = (ways + dp(pos+1, step+1))%kMod;
        }
        
        memo[pos][step] = ways;
        return ways;
    }
    
    int numWays(int steps, int arrLen) {
        this->arrLen = min(arrLen, steps);
        memo = vector(this->arrLen, vector(steps+1, -1));   // using this->memo is also OK
        this->steps = steps;
        return dp(0, 0);
    }
};

File: 1269-Number of Ways to Stay in the Same Place After Some Steps.py
class Solution:
    def numWays(self, steps: int, arrLen: int) -> int:
        # dp[steps taken][current position] = ways to come to this state
        arrLen = min(arrLen, steps+1)
        dp = [ [0]*arrLen for _ in range(steps+1) ]
        dp[0][0] = 1
        kMod = 10**9+7
        
        for step in range(1, steps+1):
            for pos in range(arrLen):
                for move in range(-1, 2):
                    if 0 <= pos + move < arrLen:
                        dp[step][pos+move] = (dp[step][pos+move] + dp[step-1][pos])%kMod
                        
        return dp[steps][0]%kMod